components/templates/src/global_fns/load_data.rs

```diff
@@ -0,0 +1,403 @@
+extern crate toml;
+extern crate serde_json;
+
+use utils::fs::{read_file, is_path_in_directory, get_file_time};
+
+use std::hash::{Hasher, Hash};
+use std::fmt;
+use std::collections::hash_map::DefaultHasher;
+use reqwest::{Client, header};
+use url::Url;
+
+use std::path::PathBuf;
+use std::sync::{Arc, Mutex};
+
+
+use csv::Reader;
+use std::collections::HashMap;
+use tera::{GlobalFn, Value, from_value, to_value, Result, Map};
+
+static GET_DATA_ARGUMENT_ERROR_MESSAGE: &str = "`load_data`: requires EITHER a `path` or `url` argument";
+
+enum DataSource {
+    Url(Url),
+    Path(PathBuf)
+}
+
+#[derive(Debug)]
+enum OutputFormat {
+    TOML,
+    JSON,
+    CSV,
+    Plain
+}
+
+impl fmt::Display for OutputFormat {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Debug::fmt(self, f)
+    }
+}
+
+impl Hash for OutputFormat {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.to_string().hash(state);
+    }
+}
+
+impl OutputFormat {
+    fn from(output_format: String) -> Result<Self> {
+        return match output_format.as_str() {
+            "toml" => Ok(OutputFormat::TOML),
+            "csv" => Ok(OutputFormat::CSV),
+            "json" => Ok(OutputFormat::JSON),
+            "plain" => Ok(OutputFormat::Plain),
+            format => Err(format!("Unknown output format {}", format).into())
+        };
+    }
+
+    fn as_accept_header(&self) -> header::HeaderValue {
+        return header::HeaderValue::from_static(match self {
+            OutputFormat::JSON => "application/json",
+            OutputFormat::CSV => "text/csv",
+            OutputFormat::TOML => "application/toml",
+            OutputFormat::Plain => "text/plain",
+        });
+    }
+}
+
+impl DataSource {
+    fn from_args(path_arg: Option<String>, url_arg: Option<String>, content_path: &PathBuf) -> Result<Self> {
+        if path_arg.is_some() && url_arg.is_some() {
+            return Err(GET_DATA_ARGUMENT_ERROR_MESSAGE.into());
+        }
+
+        if let Some(path) = path_arg {
+            let full_path = content_path.join(path);
+            if !full_path.exists() {
+                return Err(format!("{} doesn't exist", full_path.display()).into());
+            }
+            return Ok(DataSource::Path(full_path));
+        }
+
+        if let Some(url) = url_arg {
+            return Url::parse(&url).map(|parsed_url| DataSource::Url(parsed_url)).map_err(|e| format!("Failed to parse {} as url: {}", url, e).into());
+        }
+
+        return Err(GET_DATA_ARGUMENT_ERROR_MESSAGE.into());
+    }
+}
+
+fn get_cache_key(data_source: &DataSource, format: &OutputFormat) -> u64 {
```

Comment:
Making this part of the impl of DataSource would be cleaner imo

Reply:
I was toying with where to put it, as it's also equally valid on `OutputFormat`,  but as most of the computation varys for `DataSource`, I guess there makes the most sense of the 2.